File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {

        stack<int> s;
        int m = nums1.size();
        int n = nums2.size();
        vector<int> ans(n);
        for (int i = n - 1; i >= 0; i--) {

            while (!s.empty() && s.top() <= nums2[i]) {
                s.pop();
            }
            if (s.empty()) {
                ans[i] = -1;
            } else {
                ans[i] = s.top();
            }
            s.push(nums2[i]);
        }
vector<int>res;
        for(int i = 0 ; i<m;i++){
            for(int j = 0 ; j<n;j++){
                if(nums1[i]==nums2[j]){
                    res.push_back(ans[j]);
                }
            }
        }
        return res;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st=deque()
        m={}

        for a in reversed(nums2):

            while st and st[-1]<=a:
                st.pop()
            if st:
                m[a]=st[-1]
            else:
                m[a]=-1
            st.append(a)
        ans=[]
        for i in range(0,len(nums1)):
            ans.append(m[nums1[i]])
        return ans